src/node.rs

```diff
@@ -371,7 +371,11 @@ impl<K, V> TreeBin<K, V> {
                         // not obtain new references to our thread handle. Also,
                         // we just swapped out that handle, so it is no longer
                         // reachable.
-                        let _ = unsafe { waiter.into_owned() };
+                        //
+                        // Now, we cannot safely drop this _immediately_, since
+                        // we may have woken up and reached here _while_ some
```

Comment:
Sorry for only getting to reviewing this after you already merged, but I think you're missing something like a "reading thread" or "one" here 😅

Reply:
Also I think once we actually park, we cannot get here before the call to unpark (so it would be only if we _intended_ to wait, but then saw that the reader was gone in the re-check before, in a context switch from after the reader saw the `WAITER` in `lock_state`). But I think the general case is clear enough.

Reply:
Good catch! Fixed in cd19e2cc88b6c10aed77b795c77f365678a3c441